File: src/epub_merge_tool/split.py
from __future__ import annotations

import json
import warnings
import zipfile
from pathlib import Path

from .epub_io import (
    MERGE_MANIFEST_PATH,
    build_nav_html,
    build_opf,
    read_source_book,
    require_manifest,
    write_epub_container,
    write_mimetype_first,
)
from .errors import EpubMergeError, ManifestError
from .models import ManifestItem, TocEntry


def split_epub(input_path: Path | str, out_dir: Path | str, *, heuristic: bool = False) -> list[Path]:
    input_path = Path(input_path).expanduser()
    out_dir = Path(out_dir).expanduser()
    out_dir.mkdir(parents=True, exist_ok=True)

    try:
        with zipfile.ZipFile(input_path, "r") as zf:
            manifest = require_manifest(zf.read(MERGE_MANIFEST_PATH))
            return _split_from_manifest(zf, manifest, out_dir)
    except KeyError as exc:
        if not heuristic:
            raise ManifestError("missing epub-merge-tool manifest; use --heuristic for best-effort split") from exc
    if heuristic:
        warnings.warn("heuristic split is best-effort and not logically lossless", UserWarning, stacklevel=2)
        source = read_source_book(input_path)
        output = out_dir / _safe_title_filename(source.title)
        output.write_bytes(input_path.read_bytes())
        return [output]
    raise EpubMergeError("unreachable split state")


def _split_from_manifest(zf: zipfile.ZipFile, manifest: dict, out_dir: Path) -> list[Path]:
    outputs: list[Path] = []
    for source in manifest["sources"]:
        basename = source["basename"]
        output = out_dir / basename
        items = [
            ManifestItem(
                item_id=file_record["id"],
                href=file_record["href"],
                media_type=file_record["media_type"],
                properties=tuple(prop for prop in file_record.get("properties", []) if prop != "nav"),
            )
            for file_record in source["files"]
        ]
        spine_ids = source["spine"]
        toc_entries = [TocEntry(entry["title"], entry["href"]) for entry in source["toc"]]
        spine_hrefs = {
            file_record["href"]: file_record["id"]
            for file_record in source["files"]
            if file_record["id"] in spine_ids
        }
        with zipfile.ZipFile(output, "w", compression=zipfile.ZIP_DEFLATED) as out:
            write_mimetype_first(out)
            write_epub_container(out)
            out.writestr(
                "OEBPS/nav.xhtml",
                build_nav_html(
                    source["title"],
                    toc_entries[0].href if toc_entries else "#",
                    [(entry.title, entry.href, []) for entry in toc_entries],
                ),
                compress_type=zipfile.ZIP_DEFLATED,
            )
            out.writestr(
                "OEBPS/content.opf",
                build_opf(
                    source["title"],
                    source.get("language") or "en",
                    source.get("creators") or (),
                    items,
                    [spine_hrefs[_href_for_id(source["files"], item_id)] for item_id in spine_ids],
                    nav_href="nav.xhtml",
                ),
                compress_type=zipfile.ZIP_DEFLATED,
            )
            rewrites = source.get("rewrites", {})
            for file_record in source["files"]:
                data = zf.read(f"OEBPS/{file_record['merged_href']}")
                reverse = {new: old for old, new in rewrites.get(file_record["href"], {}).items()}
                if reverse:
                    data = _rewrite_refs(data, reverse)
                out.writestr(f"OEBPS/{file_record['href']}", data, compress_type=zipfile.ZIP_DEFLATED)
        outputs.append(output)
    return outputs


def _href_for_id(files: list[dict], item_id: str) -> str:
    for file_record in files:
        if file_record["id"] == item_id:
            return file_record["href"]
    raise EpubMergeError(f"manifest spine references missing file id {item_id!r}")


def _rewrite_refs(data: bytes, mapping: dict[str, str]) -> bytes:
    text = data.decode("utf-8")
    for old, new in mapping.items():
        text = text.replace(f'src="{old}"', f'src="{new}"')
        text = text.replace(f"src='{old}'", f"src='{new}'")
        text = text.replace(f'href="{old}"', f'href="{new}"')
        text = text.replace(f"href='{old}'", f"href='{new}'")
    return text.encode("utf-8")


def _safe_title_filename(title: str) -> str:
    safe = "".join(ch if ch.isalnum() or ch in {" ", ".", "-", "_"} else "_" for ch in title).strip()
    if not safe:
        safe = "split"
    return f"{safe}.epub"

File: ts/src/split.ts
import { ManifestError } from "./errors.js";
import { baseZipEntries, buildOpf, buildVolumeNav, MERGE_MANIFEST_PATH, requireMergeManifest } from "./epub-io.js";
import type { SplitResult, TocEntry, ManifestItem } from "./models.js";
import { utf8Decode, utf8Encode } from "./utils.js";
import { readZip, writeZip, type ZipEntry } from "./zip.js";

export function splitEpub(input: Uint8Array): SplitResult[] {
  const zip = readZip(input);
  const manifestData = zip.get(MERGE_MANIFEST_PATH);
  if (!manifestData) throw new ManifestError("missing epub-merge-tool manifest; use heuristic split for best-effort split");
  const manifest = requireMergeManifest(manifestData);
  const sources = manifest.sources as Array<Record<string, unknown>>;
  return sources.map((source) => splitSource(zip, source));
}

function splitSource(mergedZip: Map<string, Uint8Array>, source: Record<string, unknown>): SplitResult {
  const basename = String(source.basename);
  const files = source.files as Array<Record<string, unknown>>;
  const spine = source.spine as string[];
  const toc = source.toc as TocEntry[];
  const rewrites = (source.rewrites ?? {}) as Record<string, Record<string, string>>;
  const items: ManifestItem[] = files.map((file) => ({
    id: String(file.id),
    href: String(file.href),
    mediaType: String(file.media_type),
    properties: (file.properties as string[] | undefined ?? []).filter((p) => p !== "nav")
  }));
  const zipEntries: ZipEntry[] = baseZipEntries();
  zipEntries.push({
    name: "OEBPS/nav.xhtml",
    data: buildVolumeNav(String(source.title), toc[0]?.href ?? "#", toc.map((entry) => ({ title: entry.title, href: entry.href, entries: [] })))
  });
  zipEntries.push({
    name: "OEBPS/content.opf",
    data: buildOpf(String(source.title), String(source.language ?? "en"), (source.creators as string[] | undefined) ?? [], items, spine, "nav.xhtml")
  });
  for (const file of files) {
    const mergedHref = String(file.merged_href);
    const originalHref = String(file.href);
    const data = mergedZip.get(`OEBPS/${mergedHref}`);
    if (!data) throw new ManifestError(`manifest references missing merged resource ${mergedHref}`);
    zipEntries.push({ name: `OEBPS/${originalHref}`, data: reverseRewrite(data, rewrites[originalHref] ?? {}) });
  }
  return { name: basename, data: writeZip(zipEntries) };
}

function reverseRewrite(data: Uint8Array, mapping: Record<string, string>): Uint8Array {
  let text = utf8Decode(data);
  for (const [oldHref, newHref] of Object.entries(mapping)) {
    text = text
      .replaceAll(`src="${newHref}"`, `src="${oldHref}"`)
      .replaceAll(`src='${newHref}'`, `src='${oldHref}'`)
      .replaceAll(`href="${newHref}"`, `href="${oldHref}"`)
      .replaceAll(`href='${newHref}'`, `href='${oldHref}'`);
  }
  return utf8Encode(text);
}
